### PR TITLE
feat: Add the authentication

### LINKED ISSUE #26: Integrate authentication.

### PR DESCRIPTION
**Pull request by harish-aka-shivi**
## Description

This PR contains two commits because this branch rooted originated `harish-aka-shivi-issue-17`. That branch has not been merged yet, so I will squash the commit once that branch is merged.
nge.

Fixes #26 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Unit Test
- [ ] Integration Test

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: backend/src/models/userModel.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+userSchema.statics.comparePassword = async function comparePassword(candidatePassword, storedHash) {
+  const isMatch = await bcrypt.compare(candidatePassword, storedHash);
+  return isMatch;
+};
+
+userSchema.methods.generateAuthToken = async function generateAuthToken() {
+  const user = this;
+  // const timeInMilliseconds = Date.now();
+  const token = jwt
+    .sign({ _id: user._id.toString() }, process.env.JSON_WEB_SECRET_KEY);
```
Comment (nimish-gupta) [inline backend/src/models/userModel.js]: Import all the env variables from a constant/config file. Check this issue https://github.com/nodejs/node/issues/26213

File: backend/src/models/userModel.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+});
+
+userSchema.statics.comparePassword = async function comparePassword(candidatePassword, storedHash) {
+  const isMatch = await bcrypt.compare(candidatePassword, storedHash);
+  return isMatch;
+};
+
+userSchema.methods.generateAuthToken = async function generateAuthToken() {
+  const user = this;
+  // const timeInMilliseconds = Date.now();
```
Comment (nimish-gupta) [inline backend/src/models/userModel.js]: Remove comments.

File: backend/src/models/userModel.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
+/* eslint-disable no-use-before-define */
```
Comment (nimish-gupta) [inline backend/src/models/userModel.js]: Remove eslint comments.

File: backend/src/middlewares/auth.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  try {
+    // console.log('in middleware')
+    const token = req.header('Authorization').replace('Bearer ', '');
+    const decoded = jwt.verify(token, process.env.JSON_WEB_SECRET_KEY);
+    // TODO: Add time factor.
+    const user = await User.findOne({ _id: decoded._id, 'tokens.token': token });
+
+    if (!user) {
+      throw new Error();
+    }
```
Comment (nimish-gupta) [inline backend/src/middlewares/auth.js]: Add new line after every block, helps in making code easier to read.

File: backend/src/middlewares/auth.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
+/* eslint-disable no-underscore-dangle */
+import jwt from 'jsonwebtoken';
+import User from '../models/userModel';
+
+const auth = async (req, res, next) => {
+  try {
+    // console.log('in middleware')
```
Comment (nimish-gupta) [inline backend/src/middlewares/auth.js]: Remove comments.

File: backend/src/controllers/userController.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    res.status(500).send();
+  }
+};
+
+const deleteUser = async (req, res) => {
+  try {
+    req.user.remove();
+    res.send(req.user);
+  } catch (error) {
+    // console.log(err)
```
Comment (nimish-gupta) [inline backend/src/controllers/userController.js]: Remove comments.

File: backend/src/controllers/userController.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  } catch (error) {
+    res.status(400).send('Error in creating user');
+  }
+};
+
+const logoutUser = async (req, res) => {
+  try {
+    // console.log(req.user.tokens)
+    req.user.tokens = req.user.tokens.filter(token => token.token !== req.token);
+    // console.log(req.user.tokens)
```
Comment (nimish-gupta) [inline backend/src/controllers/userController.js]: Remove comments.

File: backend/src/controllers/userController.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    const token = await user.generateAuthToken();
+    res.send({ user, token });
+  } catch (error) {
+    res.status(400).send('Error in creating user');
+  }
+};
+
+const logoutUser = async (req, res) => {
+  try {
+    // console.log(req.user.tokens)
```
Comment (nimish-gupta) [inline backend/src/controllers/userController.js]: Remove comments.

File: backend/src/controllers/userController.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    // console.log(user);
+    const token = await user.generateAuthToken();
+    res.send({ user, token });
+  } catch (error) {
+    res.status(500).send('Error in login api');
+  }
+};
+
+const userCreate = async (req, res) => {
+  // console.log(req.body, User);
```
Comment (nimish-gupta) [inline backend/src/controllers/userController.js]: Remove comments.

File: backend/src/controllers/userController.js [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    await req.user.save();
+    res.send();
+  } catch (error) {
+    res.status(500).send();
+  }
+};
+
+const deleteUser = async (req, res) => {
+  try {
+    req.user.remove();
```
Comment (nimish-gupta) [inline backend/src/controllers/userController.js]: This is a async request right?

File: backend/src/models/userModel.js (hunk 1 above)
Comment (harish-aka-shivi) [reply backend/src/models/userModel.js]: Done, now have a config file and environment variables are assigned once to const variables.

File: backend/src/models/userModel.js (hunk 3 above)
Comment (harish-aka-shivi) [reply backend/src/models/userModel.js]: Fixed, It was accessing User Model before Modes was instantiated, now I am accessing User using this

File: backend/src/controllers/userController.js (hunk 10 above)
Comment (harish-aka-shivi) [reply backend/src/controllers/userController.js]: Agreed, remove the async from the function.👍

File: backend/src/controllers/userController.js (hunk 9 above)
Comment (harish-aka-shivi) [reply backend/src/controllers/userController.js]: Fixed

**Pull request merged**